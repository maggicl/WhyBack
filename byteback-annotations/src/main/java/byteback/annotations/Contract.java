package byteback.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Annotations for the specification of contracts.
 */
public interface Contract {

    /**
     * Declares a predicate method.
     *
     * A predicate method is a completely pure static method containing boolean
     * statements verifying conditions on its inputs. A predicate method may only
     * indirectly call other static predicate methods.
     *
     * There could possibly be exceptions for certain virtual or interface calls
     * needed for the specification of quantifiers.
     *
     * TODO: delineate such exceptions here.
     */
    @Retention(RetentionPolicy.RUNTIME)
    @Target({ ElementType.METHOD })
    public static @interface Predicate {
    }

    /**
     * Declares a condition method.
     *
     * Condition methods can be used to represent preconditions, postconditions or
     * invariants. Every condition method must be both, static and pure. The return
     * type of the annotated method must be void, as conditions are enforced by
     * means of assertions. A condition method may only call predicate methods.
     */
    @Retention(RetentionPolicy.RUNTIME)
    @Target({ ElementType.METHOD })
    public static @interface Condition {

        /**
         * Declares a parameter containing the old invocation target.
         */
        @Retention(RetentionPolicy.RUNTIME)
        @Target({ ElementType.PARAMETER })
        public static @interface Old {
        }

        /**
         * Declares a parameter containing the current invocation target.
         */
        @Retention(RetentionPolicy.RUNTIME)
        @Target({ ElementType.PARAMETER })
        public static @interface Current {
        }

        /**
         * Declares a parameter containing the return value of the invocation.
         */
        @Retention(RetentionPolicy.RUNTIME)
        @Target({ ElementType.PARAMETER })
        public static @interface Returns {
        }

    }

}
