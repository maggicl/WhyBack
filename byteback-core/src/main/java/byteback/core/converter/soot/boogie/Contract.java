package byteback.core.converter.soot.boogie;

public class Contract {

	public static String PRELUDE_ANNOTATION = "Lbyteback/annotations/Contract$Prelude;";

	public static String PURE_ANNOTATION = "Lbyteback/annotations/Contract$Pure;";

	public static String REQUIRE_ANNOTATION = "Lbyteback/annotations/Contract$Require;";

	public static String ENSURE_ANNOTATION = "Lbyteback/annotations/Contract$Ensure;";

}
