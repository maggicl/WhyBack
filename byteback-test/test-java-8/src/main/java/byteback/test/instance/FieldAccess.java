/**
 * RUN: %{byteback} -cp %{jar} -c byteback.test.instance.FieldAccess -o %t.mlw
 */
package byteback.test.instance;

import byteback.annotations.Contract.Pure;

public class FieldAccess {

	public static int staticField;

	public int field;

	@Pure
	public int staticFieldReference() {
		return staticField;
	}

	@Pure
	public int staticFieldSum() {
		return staticField + 2;
	}

	@Pure
	public int fieldReference() {
		return this.field;
	}

	@Pure
	public int fieldSum() {
		return this.field + 2;
	}

}
/**
 * RUN: %{verify} %t.mlw
 * CHECK-IGNORE: Boogie program verifier finished with 1 verified, 0 errors
 */
