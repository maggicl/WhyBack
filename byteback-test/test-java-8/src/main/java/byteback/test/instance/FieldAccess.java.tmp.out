anomaly: Sys_error("test-java-8/src/main/java/byteback/test/instance/FieldAccess.java.tmp.mlcfg: No such file or directory")
